### PR TITLE
Solution to  Memory leaks by interning external strings #24 per @sletner

### PR DESCRIPTION
**Pull request by daninus14**

### CODE REVIEW

**Comment by daninus14** [comment]
This addresses this issue: https://github.com/edicl/hunchentoot/issues/24

File: headers.lisp [hunk 1]
```diff
@@ -26,6 +26,22 @@
 ;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 ;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+;;; The following code is for treating headers which exist in the keyword package as keywords, otherwise they will be treated as strings. This code will override the chunga code which by default interns every header to be a keyword
+
+(in-package :chunga)
+
+(defvar *intern-unsafely*)
```
Comment (hanshuebner) [inline headers.lisp:33]: Initialize this with `nil` to simplify the check in `as-keyword`.

File: headers.lisp [hunk 2]
```diff
@@ -26,6 +26,22 @@
...
+
+(defun as-keyword (string &key (destructivep t))
+  "Checks if the string STRING is found as a keyword and if it is, returns the keyword, otherwise it returns a string. Note that this is obviously not congruent to the name of the function, it is meant to be an override to the default behavior to avoid memory leaks in Hunchentoot"
+  (or (gethash string +string-to-keyword-hash+)
```
Comment (hanshuebner) [inline headers.lisp:37]: I would not go through `+string-to-keyword-hash+` and instead invoke `find-symbol` directly.

File: headers.lisp [hunk 3]
```diff
@@ -26,6 +26,22 @@
...
+      (find-symbol (string-upcase string) (find-package "KEYWORD"))
+      (if (and (boundp '*intern-unsafely*) *intern-unsafely*)
```
Comment (hanshuebner) [inline headers.lisp:39]: Simplify to
```suggestion
      (if *intern-unsafely*
```

**Review by hanshuebner — CHANGES_REQUESTED** [review]
I just realized that you're monkey-patching `CHUNGA`, which I'm not a fan of.  I think it would be preferable to fix the issue there.

**Comment by daninus14** [comment]
> I just realized that you're monkey-patching `CHUNGA`, which I'm not a fan of. I think it would be preferable to fix the issue there.

So should I make this changes in chunga instead and open a PR there?

File: headers.lisp (hunk 1 above)
Comment (daninus14) [reply headers.lisp]: just added it

File: headers.lisp (hunk 2 above)
Comment (daninus14) [reply headers.lisp]: just added it

File: headers.lisp (hunk 3 above)
Comment (daninus14) [reply headers.lisp]: just added it

**Comment by daninus14** [comment]
> > > I just realized that you're monkey-patching `CHUNGA`, which I'm not a fan of. I think it would be preferable to fix the issue there.
> > 
> > 
> > So should I make this changes in chunga instead and open a PR there?
> 
> Yes please!

Ok, so I made a PR there, will make changes here to account for that as well

**Comment by daninus14** [comment]
Ok, made the [PR over there](https://github.com/edicl/chunga/pull/11), please take a look at the code since I am implementing it differently than here since it's all internal to chunga and there's no need for the `*intern-unsafely*`

**Comment by hanshuebner** [comment]
The `CHUNGA` PR is merged, thank you!

**Comment by daninus14** [comment]
Thank you as well for the guidance and being so responsive!

**Pull request closed without merging**